cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringProgramEventPublisher.java

```diff
@@ -0,0 +1,232 @@
+                                              lastProgramUpdateMessageId);
+
+    });
+
+    return pollInterval;
+  }
+
+  /**
+   * Returns program status updates for tethered peers.
+   */
+  private PeerProgramUpdates getPeerProgramUpdates() {
+    Map<String, List<Notification>> peerToNotifications = new HashMap<>();
+    String lastMessageId = lastProgramUpdateMessageId;
+    try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                    programUpdateTopic,
+                                                                    programUpdateFetchSize,
+                                                                    lastProgramUpdateMessageId)) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        Notification notification = message.decodePayload(r -> GSON.fromJson(r, Notification.class));
+        if (notification.getNotificationType() == Notification.Type.PROGRAM_STATUS) {
+          Map<String, String> properties = notification.getProperties();
+          String programRunId = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
+          try {
+            RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
+            if (runRecord.getPeerName() != null) {
+              String programStatus = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
+              LOG.debug("Notifying peer {} about program run {} in state {}",
+                        runRecord.getPeerName(), programRunId, programStatus);
+              peerToNotifications.computeIfAbsent(runRecord.getPeerName(), n -> new ArrayList<>()).add(notification);
+            }
+          } catch (NotFoundException | IOException e) {
+            LOG.error("Unable to fetch runRecord for programRunId {}", programRunId, e);
+          }
+        }
+        lastMessageId = message.getId();
+      }
+    } catch (Exception e) {
+      LOG.error("Exception when fetching program updates. Will retry again during next poll", e);
+    }
+    return new PeerProgramUpdates(peerToNotifications, lastMessageId);
+  }
+
+  private void persistNotifications(PeerProgramUpdates peerProgramUpdates, List<PeerInfo> peers) throws IOException {
+    for (PeerInfo peer: peers) {
+      String peerName = peer.getName();
+      List<String> notifications = peerProgramUpdates.peerToNotifications.getOrDefault(peerName, new ArrayList<>())
+        .stream().map(n -> GSON.toJson(n)).collect(Collectors.toList());
+      publishMessages(programStateTopicPrefix + peerName, notifications);
+    }
+    lastProgramUpdateMessageId = peerProgramUpdates.lastMessageId;
+  }
+
+  private void publishMessages(String topic, List<String> messages) throws IOException {
+    if (messages.isEmpty()) {
+      return;
+    }
+    int retries = 0;
+    while (++retries < 2) {
```

Comment:
retries should be done with some time in between and with backoff (for example, using the Retries utility class). Otherwise you can have cascading types of failures, where a failure in one place triggers more widespread failures as retries overload the system.

Reply:
Now using Retries.runWithRetries() with 1 retry